### PR TITLE
Add a helper model to automatically apply preprocessing

### PR DESCRIPTION
**Pull request by mattdangerw**

### CODE REVIEW

File: keras_nlp/utils/pipeline_model.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    x=None,
+    y=None,
+    sample_weight=None,
+    batch_size=None,
+):
+    if isinstance(x, tf.data.Dataset):
+        if y is not None:
+            raise ValueError(
+                "When `x` is a `tf.data.Dataset`, please do not provide "
+                f"`y`. Received: `y={y}`."
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:31]: Might have to truncate or reformat the `y` display because if it's an array that will be a very unreadable message

File: keras_nlp/utils/pipeline_model.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            )
+        if sample_weight is not None:
+            raise ValueError(
+                "When `x` is a `tf.data.Dataset`, please do not provide "
+                f"`sample_weight`. Received: `sample_weight={sample_weight}`."
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:36]: Same comment. Maybe just display the type?

File: keras_nlp/utils/pipeline_model.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    inputs = keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
+    return tf.data.Dataset.from_tensor_slices(inputs).batch(batch_size or 32)
+
+
+def apply_dataset_fn(ds, fn):
+    return ds.map(fn, tf.data.AUTOTUNE).prefetch(tf.data.AUTOTUNE)
+
+
+class PipelineModel(keras.Model):
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:48]: Good name

File: keras_nlp/utils/pipeline_model.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    def preprocess_features(self, x):
+        """An overridable function which preprocesses features."""
+        return x
+
+    def preprocess_labels(self, y):
+        """An overridable function which preprocesses labels."""
+        return y
+
+    def preprocess(self, *args):
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:63]: I wonder if we should make it more specific? `preprocess_data`? or is that confusing

File: keras_nlp/utils/pipeline_model.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        sample_weight=None,
+        validation_data=None,
+        validation_split=None,
+        **kwargs,
+    ):
+        x = convert_inputs_to_dataset(x, y, sample_weight, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess)
+
+        if validation_split:
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:86]: We should support it. It's straightforward -- see RetinaNet example

File: keras_nlp/utils/pipeline_model.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            raise ValueError(
+                "`validation_split` is not supported on this model. "
+                f"Received: `validation_split={validation_split}`."
+            )
+        if validation_data is not None:
+            val_tuple = keras.utils.unpack_x_y_sample_weight(validation_data)
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:92]: `x, y, sw` may be more readable

File: keras_nlp/utils/pipeline_model.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                "When `x` is a `tf.data.Dataset`, please do not provide "
+                f"`sample_weight`. Received: `sample_weight={sample_weight}`."
+            )
+        return x
+
+    inputs = keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
+    return tf.data.Dataset.from_tensor_slices(inputs).batch(batch_size or 32)
+
+
+def apply_dataset_fn(ds, fn):
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:44]: `map_fn_and_prefetch` is more explicit -- but I would actually suggest not having this function at all, just inline to reduce indirection

File: keras_nlp/utils/pipeline_model.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+
+class PipelineModel(keras.Model):
+    """A model which allows automatically applying preprocessing."""
+
+    def __init__(self, *args, include_preprocessing=True, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.include_preprocessing = include_preprocessing
+
+    def preprocess_features(self, x):
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:55]: In a fully generalizable API we are likely going to need a `training` arg here and below

**Review by fchollet — COMMENTED** [review]
Looks great!

File: keras_nlp/utils/pipeline_model.py (hunk 5 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py:86]: Ah yeah, I think there are some bugs in the CV version. We can add this, just we need a little more complexity than what's there.

[These lines](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/object_detection/object_detection_base_model.py#L47-L53) actually keep the validation data in the training data! Pretty broken I think.

Will ping the CV folk.

File: keras_nlp/utils/pipeline_model.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+        return super().evaluate(
+            x=x,
+            y=None,
+            batch_size=None,
+            sample_weight=None,
+            **kwargs,
+        )
+
+    def predict(
```
Comment (mattdangerw) [inline keras_nlp/utils/pipeline_model.py:128]: Some performance optimization? Converting to a dataset in the single sample case will be inefficient.

Add a check. If input is not a dataset and is short, avoid extra work here.

File: keras_nlp/utils/pipeline_model.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        x = convert_inputs_to_dataset(x, None, None, batch_size)
+        if self.include_preprocessing:
+            x = apply_dataset_fn(x, self.preprocess_features)
+
+        return super().predict(
+            x=x,
+            batch_size=None,
+            **kwargs,
+        )
+
```
Comment (mattdangerw) [inline keras_nlp/utils/pipeline_model.py:143]: Also add train_on_batch, test_on_batch, predict_on_batch. These don't work with datasets, so things should get simpler.

File: keras_nlp/utils/pipeline_model.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    def __call__(self, inputs, include_preprocessing=None, **kwargs):
+        if include_preprocessing is None:
```
Comment (mattdangerw) [inline keras_nlp/utils/pipeline_model.py:145]: Let's try just making this dtype aware to avoid preprocessing what has already been preprocessed.

**Comment by jbischof** [comment]
This is a cool idea very similar to Huggingface pipelines. Is there any efficiency loss in taking this "lazy" approach to tokenization? Are all the optimizations described in the tfds guide still possible ([doc](https://www.tensorflow.org/guide/data_performance))?

HF docs warn users to expect a performance hit for batched data when there is high variability in sequence lengths ([doc](https://huggingface.co/docs/transformers/v4.21.3/en/main_classes/pipelines#pipeline-batching)).

**Comment by mattdangerw** [comment]
Looking at there summary of performance optimization, I would say largely yes, with a few caveats.

 - We are not loading the data, so sharding and interleaving data on the loading side is still up to the user.
 - We are not caching the data in memory, as that is not good general purpose approach (it might not fit in memory, there might be random augmentations). We could consider an argument, though I would be inclined to not add it.
 - Some people may want to preprocess in a separate job entirely, which will often be the fastest, at an increased usability burden and an inability to do things like generate random masks. This is clearly out of scope for what we can do in a helper like this, though still fully supported with `include_preprocessing=False`

I would actually expect this preprocessing to do better than the hf approach (unless they are now leveraging tf.data under the hood?), but we would need to validate this.

The question of whether preprocessing might be better applied batched or unbatched is interesting. But I don't think we need to be everything for everyone with this, as you can toggle it on and off. If it is always better to preprocess unbatched vs batched, this design could be worrisome. But I don't believe that is the case.

tl;dr for most users, this is a speedup, where we correctly apply data preprocessing tricks they aren't aware of. For power users where we aren't doing things exactly as they want, just pass `include_preprocessing=False`.

File: keras_nlp/utils/pipeline_model.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+def convert_inputs_to_dataset(
+    x=None,
+    y=None,
+    sample_weight=None,
+    batch_size=None,
```
Comment (jbischof) [inline keras_nlp/utils/pipeline_model.py:25]: why does this not default to 32?

File: keras_nlp/utils/pipeline_model.py (hunk 1 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py:31]: Is there a good way to do this that's safe with a wide range of input types? Just removing for now.

Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py:31]: Ah seeing comment below, I will print the type!

File: keras_nlp/utils/pipeline_model.py (hunk 8 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py:107]: I think this might be easier to do when upstreaming into core Keras. Getting `training` correct at call time seems to rely on some internal stuff, and we don't really need it on the NLP side so much I think.

https://github.com/keras-team/keras/blob/master/keras/engine/base_layer.py#L1056

File: keras_nlp/utils/pipeline_model.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+
+        return super().predict(
+            x=x,
+            batch_size=None,
+            **kwargs,
+        )
+
+    def __call__(self, inputs, include_preprocessing=None, **kwargs):
+        # We don't trace if `including_preprocessing` is `False`, or we are
+        # currently tracing a functional model.
```
Comment (fchollet) [inline keras_nlp/utils/pipeline_model.py:206]: What does "trace" mean in this context? I'm confused.

File: keras_nlp/utils/pipeline_model.py (hunk 3 above)
Comment (fchollet) [reply keras_nlp/utils/pipeline_model.py:102]: It's also a very well known sklearn class that chains preprocessing and a model: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html

**Review by fchollet — APPROVED** [review]
The API looks great! LGTM from me.

File: keras_nlp/utils/pipeline_model.py (hunk 13 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py:206]: Hmm, yeah maybe the wrong word. Wanted to avoid attempting to apply preprocessing when we are constructing a graph with functional model inputs, because applying preprocessing to functional inputs would never work.

File: keras_nlp/utils/pipeline_model.py [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+
+    def evaluate(
+        self,
+        x=None,
+        y=None,
+        batch_size=None,
+        sample_weight=None,
+        **kwargs,
+    ):
+        # TODO: we can't support a cached eval dataset until we make changes to
```
Comment (jbischof) [inline keras_nlp/utils/pipeline_model.py:177]: Should we issue a warning?

File: keras_nlp/utils/pipeline_model.py (hunk 9 above)
Comment (jbischof) [reply keras_nlp/utils/pipeline_model.py:192]: I assume we could have a simple list comprehension or python `map` when the input length < x?

Comment (jbischof) [reply keras_nlp/utils/pipeline_model.py:192]: Where `x = tf.data.AUTOTUNE` lol

File: keras_nlp/utils/pipeline_model.py [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+        )
+
+    def __call__(self, inputs, include_preprocessing=None, **kwargs):
+        # We don't trace if `including_preprocessing` is `False`, or we are
+        # called on functional model inputs.
+        flat_inputs = tf.nest.flatten(inputs)
+        is_functional_input = any(
+            [type(t).__name__ == "KerasTensor" for t in flat_inputs]
+        )
+        if include_preprocessing is None:
```
Comment (jbischof) [inline keras_nlp/utils/pipeline_model.py:287]: Move these two lines to the top for clarity

File: keras_nlp/utils/pipeline_model.py [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+    ):
+        if self.include_preprocessing:
+            data = self.preprocess_samples(x)
+            x, _, _ = tf.keras.utils.unpack_x_y_sample_weight(data)
+        return super().predict_on_batch(
+            x=x,
+            **kwargs,
+        )
+
+    def __call__(self, inputs, include_preprocessing=None, **kwargs):
```
Comment (jbischof) [inline keras_nlp/utils/pipeline_model.py:280]: I don't love magic `None` values if we can avoid it. Would it be possible to have a vanilla `__call__` that respects the class default like the other methods and a special method that always skips preprocessing for internal use? Is there any reason for this override to be user facing?

**Review by jbischof — COMMENTED** [review]
Some more small comments. Let's talk soon!

File: keras_nlp/utils/pipeline_model.py (hunk 14 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py:181]: Updated the comment with a little more color.

I don't think we could. This is not something that is exposed to users. We would just be warning every `model.fit()`.

Core Keras does a weird thing during `fit()` where the `x` and `y` arguments to evaluate are ignored, and a secret property containing the validation data is read off. We could maybe rely on this mechanism as is if we apply the preprocessing to our validation data during `fit()`, but then if this caching changed in the future we would break. Felt safest to just disable it.

File: keras_nlp/utils/pipeline_model.py (hunk 9 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py:196]: Yeah, we should check with @fchollet on this, he brought this up originally. I think there are some optimizations for this in core Keras we should mirror.

File: keras_nlp/utils/pipeline_model.py (hunk 15 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py]: this is removed

File: keras_nlp/utils/pipeline_model.py (hunk 16 above)
Comment (mattdangerw) [reply keras_nlp/utils/pipeline_model.py]: (talked offline) we are going to remove all the `__call__` stuff for now, so none of this applies!

**Comment by mattdangerw** [comment]
> We could also consider never preprocessing in model.call. This might feel more inconsistent to a user, but would cut down a lot of code here for sure. And maybe the fact the preprocessing is only applied in fit/predict/evaluate would actually be a simpler mental model? Worth discussing!

After discussion, we decided to go with the more minimal version that does not override call.
